plugins/hls-change-type-signature-plugin/src/Ide/Plugin/ChangeTypeSignature.hs

```diff
@@ -0,0 +1,237 @@
+    (rss, Just ghcSig) <- listToMaybe [ (locA rss, sigToText ts)
+                                        | L (RealSrcSpan rss _) (SigD _ ts@(TypeSig _ idsSig _)) <- decls,
+                                        any ((==) declName . occNameString . occName . unLoc) idsSig
+                                      ]
+    pure (rss, ghcSig)
+
+-- | Pretty Print the Type Signature (to validate GHC Error Message)
+sigToText :: Sig GhcPs -> Maybe Text
+sigToText = \case
+  ts@TypeSig {} -> stripSignature (T.pack $ showSDocUnsafe $ ppr ts)
+  _             -> Nothing
+
+stripSignature :: Text -> Maybe Text
+-- for whatever reason incoming signatures MAY have new lines after "::" or "=>"
+stripSignature sig = case T.filter (/= '\n') sig =~ sigRegex :: (Text, Text, Text, [Text]) of
+                        -- No constraints (Monad m =>)
+                         (_, _, _, [sig'])    -> Just $ T.strip sig'
+                        -- Ignore constraints (Monad m =>)
+                         (_, _, _, [_, sig']) -> Just $ T.strip sig'
+                         _                    -> Nothing
+    where
+        -- we want to test everthing after the constraints (GHC never gives us the constraint in the expected signature)
+        sigRegex = ".* :: (.*=>)?(.*)" :: Text
+
+
+changeSigToCodeAction :: ChangeSignature -> Command |? CodeAction
+changeSigToCodeAction ChangeSignature{..} = InR CodeAction { _title       = mkChangeSigTitle declName actualType
+                                                           , _kind        = Just CodeActionQuickFix
+                                                           , _diagnostics = Just $ List [diagnostic]
+                                                           , _isPreferred = Nothing
+                                                           , _disabled    = Nothing
+                                                           , _edit        = Just $ mkChangeSigEdit uri declSrcSpan (mkNewSignature declName actualType)
+                                                           , _command     = Nothing
+                                                           , _xdata       = Nothing
+                                                           }
+
+mkChangeSigTitle :: Text -> Text -> Text
+mkChangeSigTitle declName actualType = "change signature for ‘" <> declName <> "’ to: " <> actualType
+
+mkChangeSigEdit :: Uri -> RealSrcSpan -> Text -> WorkspaceEdit
+mkChangeSigEdit uri ss replacement =
+        let txtEdit = TextEdit (realSrcSpanToRange ss) replacement
+            changes = Just $ Map.singleton uri (List [txtEdit])
+        in WorkspaceEdit changes Nothing Nothing
+
+mkNewSignature :: Text -> Text -> Text
+mkNewSignature declName actualType = declName <> " :: " <> tidyActualType actualType
+
+-- | Cleans up type signatures with oddly-named type variables (a0, m0 ...)
+tidyActualType :: Text -> Text
+tidyActualType actualType =
```

Comment:
Is this actually more complex than it's worth?

Reply:
yes